### PR TITLE
Speed up tokenizing of a row in csv and xstrtod parsing

### PR DESCRIPTION
**Pull request by vnlitvinov**
- [x] closes: N/A
- [x] tests passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

I will update the PR when CI finishes running, as I locally tested `io.parser.test_common` only.

### CODE REVIEW

File: doc/source/whatsnew/v0.25.0.rst [hunk 1]
```diff
@@ -132,7 +132,8 @@ Performance Improvements
 - Improved performance of :meth:`Series.searchsorted`. The speedup is especially large when the dtype is
   int8/int16/int32 and the searched key is within the integer bounds for the dtype (:issue:`22034`)
 - Improved performance of :meth:`pandas.core.groupby.GroupBy.quantile` (:issue:`20405`)
-
+- Improved performance of `tokenize_bytes` in `tokenizer.c`
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: also say ``:meth:`read_csv` `` a user has no idea what any of the other things are

**Review by jreback — CHANGES_REQUESTED** [review]
can you run the asv benchmarks for csv and report the changes. we *may* not have benchmarks that are specificying hitting the code you changes; if so can you add one.

**Comment by vnlitvin** [comment]
`asv continuous -f 1.05 origin/master HEAD -b io.csv` results:

before           |after         |ratio|test name
-|-|-|-
[e8d951d2]|       [a4f6dcdb]||
**master**|       **speed-up-tokenizer**||
1.58±0.02ms|      1.49±0.02ms|     0.94|  io.csv.ReadCSVFloatPrecision.time_read_csv(',', '.', 'high')
5.23±0.1ms|      4.87±0.03ms|     0.93|  io.csv.ReadUint64Integers.time_read_uint64_na_values
10.1±0.2ms|       9.15±0.3ms|     0.91|  io.csv.ReadCSVSkipRows.time_skipprows(10000)
13.0±0.3ms|       11.6±0.2ms|     0.89|  io.csv.ReadCSVThousands.time_thousands('\|', None)

So this basically gives from 5% to 10% increase on mini-benchmarks. Note that this scales quite well, and on cases where csv files are million lines long it still gives same 5-10%.

**Comment by vnlitvin** [comment]
The results are even more promising if I allow more warmup and more sampling time so Turboboost and frequency scaling don't impact the performance too much.

Running `asv continuous -f 1.05 origin/master HEAD -b io.csv -a sample_time=2 -a warmup_time=2` yields:

before           |after         |ratio|test name
-|-|-|-
[e8d951d2]       |[a4f6dcdb]||
**master**|       **speed-up-tokenizer**||
34.9±0.2ms|         32.9±1ms|     0.94|  io.csv.ReadCSVCategorical.time_convert_direct
13.5±0.02ms|      12.6±0.08ms|     0.93|  io.csv.ReadCSVThousands.time_thousands(',', ',')
5.10±0.07ms|      4.71±0.09ms|     0.92|  io.csv.ReadUint64Integers.time_read_uint64_neg_values
14.6±0.06ms|      13.0±0.04ms|     0.89|  io.csv.ReadCSVThousands.time_thousands('\|', ',')
16.0±0.3ms|      13.8±0.09ms|     0.86|  io.csv.ReadCSVSkipRows.time_skipprows(None)
10.3±0.1ms|       8.81±0.1ms|     0.86|  io.csv.ReadCSVSkipRows.time_skipprows(10000)
12.9±0.1ms|      10.7±0.09ms|     0.84|  io.csv.ReadCSVThousands.time_thousands('\|', None)
13.0±0.05ms|      10.8±0.08ms|     0.83|  io.csv.ReadCSVThousands.time_thousands(',', None)

**Comment by vnlitvin** [comment]
@jreback I've fixed whatsnew entry per your comment and rebased to latest master for clean history.

File: doc/source/whatsnew/v0.25.0.rst [hunk 2]
```diff
@@ -169,7 +169,7 @@ Performance Improvements
 - Improved performance of :meth:`Series.searchsorted`. The speedup is especially large when the dtype is
   int8/int16/int32 and the searched key is within the integer bounds for the dtype (:issue:`22034`)
 - Improved performance of :meth:`pandas.core.groupby.GroupBy.quantile` (:issue:`20405`)
-
+- Improved performance of :meth:`read_csv` (faster tokenizing and faster parsing of small float numbers)
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: can you add this issue number (the PR number as there is no issue). 

Improved performance of :meth:`read_csv` by faster tokenizing and faster parsing of small float numbers. (:issue:`....`)

**Review by jreback — CHANGES_REQUESTED** [review]
comment on whatsnew, ping on green.

File: doc/source/whatsnew/v0.25.0.rst (hunk 2 above)
Comment (vnlitvin) [reply doc/source/whatsnew/v0.25.0.rst]: Fixed

**Comment by jreback** [comment]
lgtm.

**Review by jreback — APPROVED** [review]

**Comment by vnlitvin** [comment]
> ping on green

@jreback all tests are green

**Comment by jreback** [comment]
thanks @vnlitvin nice patch!

**Pull request merged**